academy/ibc-dev/ibcmw-develop.md

```diff
@@ -0,0 +1,460 @@
+
+### Capabilities
+
+Earlier we saw that the handshake callbacks passed the capability in the callback arguments along to the underlying application so that it may be claimed by the base application. In the `ICS4Wrapper` methods, the base application will then pass the capability up the stack in order to authenticate an outgoing packet or acknowledgement.
+
+If the middleware wishes to send a packet or acknowledgment without the involvement of the underlying application, it should be given access to the same `scopedKeeper` as the base application so that it can retrieve the capabilities by itself.
+
+### Code snippets
+
+Note that the code snippets below contain _pseudo code_, like `doCustomLogic(args)`. Every code snippet is accompanied by a reference to the respective function in the [fee middleware IBC application](https://github.com/cosmos/ibc-go/tree/main/modules/apps/29-fee) from the official `ibc-go` repository.
+
+<ExpansionPanel title="`SendPacket`">
+
+```go
+func SendPacket(
+    ctx sdk.Context,
+    chanCap *capabilitytypes.Capability,
+    appPacket exported.PacketI,
+) {
+    // middleware may modify packet
+    packet = doCustomLogic(appPacket)
+    return ics4Keeper.SendPacket(ctx, chanCap, packet)
+}
+```
+
+See [here](https://github.com/cosmos/ibc-go/blob/48a6ae512b4ea42c29fdf6c6f5363f50645591a2/modules/apps/29-fee/ibc_middleware.go#L336-L343) for an example implementation of this function for the ICS29 Fee Middleware module.
+
+</ExpansionPanel>
+
+<ExpansionPanel title="`WriteAcknowledgement`">
+
+```go
+// only called for async acks
+func WriteAcknowledgement(
+    ctx sdk.Context,
+    chanCap *capabilitytypes.Capability,
+    packet exported.PacketI,
+    ack exported.Acknowledgement,
+) {
+    // middleware may modify acknowledgement
+    ack_bytes = doCustomLogic(ack)
+    return ics4Keeper.WriteAcknowledgement(packet, ack_bytes)
+}
+```
+
+See [here](https://github.com/cosmos/ibc-go/blob/48a6ae512b4ea42c29fdf6c6f5363f50645591a2/modules/apps/29-fee/ibc_middleware.go#L345-L353) for an example implementation of this function for the ICS29 Fee Middleware module.
+
+</ExpansionPanel>
+
+<ExpansionPanel title="`GetAppVersion`">
+
+```go
+// middleware must return the underlying application version
+func GetAppVersion(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+) (string, bool) {
+    version, found := ics4Keeper.GetAppVersion(ctx, portID, channelID)
+    if !found {
+        return "", false
+    }
+    if !MiddlewareEnabled {
+        return version, true
+    }
+    // unwrap channel version
+    metadata, err := Unmarshal(version)
+    if err != nil {
+        panic(fmt.Errof("unable to unmarshal version: %w", err))
+    }
+    return metadata.AppVersion, true
+}
+```
+
+</ExpansionPanel>
```

Comment:
Is the file complete? It ends abruptly, and I was expecting to see another text reading _"See [here](link) for an example..."_ after this final ExpansionPanel.

Reply:
It's not missing anything but you make a fair point, I should add some glue text here!

Reply:
:-) I did wonder if there should be some sort of wrapping up or section summary paragraph, also on the Integrate doc.